net/server.go

```diff
@@ -164,6 +207,108 @@ func (s *server) PushLog(ctx context.Context, req *pb.PushLogRequest) (*pb.PushL
 	return &pb.PushLogReply{}, nil
 }
 
+func (s *server) getEncryptionKeys(
+	ctx context.Context,
+	req *pb.FetchEncryptionKeyRequest,
+) ([]byte, []*pb.EncryptionKeyTarget, error) {
+	encryptionKeys := make([]byte, 0)
+	targets := make([]*pb.EncryptionKeyTarget, 0)
+	for _, target := range req.Targets {
+		docID, err := client.NewDocIDFromString(string(target.DocID))
+		if err != nil {
+			return nil, nil, err
+		}
+
+		optFieldName := immutable.None[string]()
+		if target.FieldName != "" {
+			optFieldName = immutable.Some(target.FieldName)
+		}
+		encKey, err := encryption.GetKey(
+			encryption.ContextWithStore(ctx, s.peer.encstore), docID.String(), optFieldName, target.Height)
+		if err != nil {
+			return nil, nil, err
+		}
+		// TODO: we should test it somehow. For this this one peer should have some keys and
+		// another one should have the others
+		if len(encKey) == 0 {
+			continue
+		}
+		targets = append(targets, target)
+		encryptionKeys = append(encryptionKeys, encKey...)
+	}
+	return encryptionKeys, targets, nil
+}
+
+func (s *server) TryGenEncryptionKey(ctx context.Context, req *pb.FetchEncryptionKeyRequest) (*pb.FetchEncryptionKeyReply, error) {
+	peerID, err := peerIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	reqPubKey := s.peer.host.Peerstore().PubKey(peerID)
+
+	isValid, err := s.verifyRequestSignature(req, reqPubKey)
+	if err != nil {
+		return nil, errors.Wrap("invalid signature", err)
+	}
+
+	if !isValid {
+		return nil, errors.New("invalid signature")
+	}
+
+	encryptionKeys, targets, err := s.getEncryptionKeys(ctx, req)
+	if err != nil || len(encryptionKeys) == 0 {
+		return nil, err
+	}
+
+	reqEphPubKey, err := crypto.X25519PublicKeyFromBytes(req.EphemeralPublicKey)
+	if err != nil {
+		return nil, errors.Wrap("failed to unmarshal ephemeral public key", err)
+	}
+
+	encryptedKey, err := crypto.EncryptECIES(encryptionKeys, reqEphPubKey, makeAssociatedData(req, s.peer.PeerID()))
+	if err != nil {
+		return nil, errors.Wrap("failed to encrypt key for requester", err)
+	}
+
+	res := &pb.FetchEncryptionKeyReply{
+		SchemaRoot:            req.SchemaRoot,
+		ReqEphemeralPublicKey: req.EphemeralPublicKey,
+		Targets:               targets,
+		EncryptedKeys:         encryptedKey,
+	}
+
+	res.Signature, err = s.signResponse(res)
```

Comment:
thought: if these signatures are only used to verify the sender is authenticate the pubsub implementation can handle this automatically https://github.com/libp2p/go-libp2p-pubsub/blob/master/sign.go

Reply:
that's a good point. @jsimnz what do you think?

Reply:
as discussed on a stand up, it's fine to leave it like this